File: 0146-lru-cache/0146-lru-cache.py
class Node: 
    def __init__(self, key, val):
        self.key = key 
        self.val = val
        self.prev = None 
        self.next = None 

class LRUCache:
    def __init__(self, capacity: int):
        self.cap = capacity 
        self.cache = {} 
        self.left = Node(0,0)
        self.right = Node(0,0)
        self.left.next = self.right 
        self.right.prev = self.left 
    
    # Removes the provided node 
    def remove(self, node):
        left = node.prev 
        right = node.next 
        left.next = right 
        right.prev = left 
        
    # Inserts the node at the end of the list (before the right node) 
    def insert(self, node):
        prevNode = self.right.prev
        self.right.prev = node 
        node.next = self.right
        node.prev = prevNode
        prevNode.next = node 
    
    def get(self, key: int) -> int:    
        if key in self.cache: 
            node = self.cache[key]
            self.remove(node)
            self.insert(node)
            return node.val 
            
        return -1 
        
    def put(self, key: int, value: int) -> None:
        if key in self.cache:
            node = self.cache[key]
            self.remove(node)
         
        node = Node(key, value)
        self.cache[key] = node 
        self.insert(node)
        
        if len(self.cache) > self.cap:
            lru = self.left.next 
            self.remove(lru)
            del self.cache[lru.key]
        

# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)

File: 0146-lru-cache/0146-lru-cache.java
import java.util.HashMap;
import java.util.Objects;

class Node {
    public int key;
    public int value;

    public Node prev;
    public Node next;

    public Node(int key, int value) {
        this.key = key;
        this.value = value;
        this.prev = null;
        this.next = null;
    }
}



class LRUCache {

    private int cap;
    private HashMap<Integer, Node> hashMap;

    private Node left;
    private Node right;
    public LRUCache(int capacity) {
        this.cap = capacity;
        this.hashMap = new HashMap<>();
        this.left = new Node(0, 0);
        this.right = new Node(0, 0);
        this.left.next = this.right;
        this.right.prev = this.left;
    }
    // Remove a particular node from the Doubly Linked List

    public void remove(Node node) {
        Node left = node.prev;
        Node right = node.next;
        left.next = right;
        right.prev = left;
    }
    // Insert a node right at the end of the list
    public void insert(Node node) {
        Node tmp = this.right.prev;
        this.right.prev = node;
        node.next = this.right;
        node.prev = tmp;
        tmp.next = node; 
    }
    public int get(int key) {

        if (hashMap.containsKey(key)) {
            Node node = this.hashMap.get(key);
            this.remove(node);
            this.insert(node);
            return node.value;
        }
        return -1;
    }

    public void put(int key, int value) {

        if (hashMap.containsKey(key)) {
            Node node = this.hashMap.get(key);
            this.remove(node);
            this.hashMap.remove(key);
        }

        Node node = new Node(key, value);
        this.hashMap.put(key, node);
        this.insert(node);

        if (hashMap.size() > this.cap) {
            Node lru = this.left.next;
            this.remove(lru);
            this.hashMap.remove(lru.key);
        }

    }
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache obj = new LRUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */